src/policy/immix/immixspace.rs

```diff
@@ -511,24 +566,36 @@ impl<VM: VMBinding> ImmixSpace<VM> {
         } else {
             // We won the forwarding race; actually forward and copy the object if it is not pinned
             // and we have sufficient space in our copy allocator
-            let new_object = if self.is_pinned(object) || self.defrag.space_exhausted() {
+            let new_object = if self.is_pinned(object)
+                || (!nursery_collection && self.defrag.space_exhausted())
+            {
                 self.attempt_mark(object, self.mark_state);
                 ForwardingWord::clear_forwarding_bits::<VM>(object);
                 Block::containing::<VM>(object).set_state(BlockState::Marked);
                 object
             } else {
-                ForwardingWord::forward_object::<VM>(object, semantics, copy_context)
+                let new_object =
+                    ForwardingWord::forward_object::<VM>(object, semantics, copy_context);
+                Block::containing::<VM>(new_object).set_state(BlockState::Marked);
+                new_object
             };
             debug_assert_eq!(
                 Block::containing::<VM>(new_object).get_state(),
                 BlockState::Marked
             );
             queue.enqueue(new_object);
             debug_assert!(new_object.is_live());
+            self.unlog_object_if_needed(new_object);
```

Comment:
A better approach is to bulk zero the unlog bits when a clean block is allocated, or when reusing a reusable hole, instead of unlog each object here. Unlog every individual object may not be faster than bulk zeroing.

Reply:
Can you elaborate on this a bit more? I am confused. Bulk zeroing is setting bits to 0, but here we set the bit to 1 (making the object a mature object). At the end of a nursery GC, all the mature objects should have bits as 1.

Also bear in mind that in sticky immix, an object becomes mature as long as we trace it in a nursery GC (we do not have to copy it to make it a mature object). So usually we unlog an object when we copy it, but in sticky immix, we need to unlog an object when we trace it.

Reply:
>  Bulk zeroing is setting bits to 0, but here we set the bit to 1 (making the object a mature object)

yes I mean bulk-memset the unlog table to 1

> So usually we unlog an object when we copy it, but in sticky immix, we need to unlog an object when we trace it.

For object left in place without moving, yes you have to unlog them without bulk-memset.

But bulk-memset can be applied to to-space. For this case, the object is already copied. The to-space memory should be bulk initialized to 1 in allocation slow-path.

Reply:
Say if you're copying 1k promoted objects to a 32 KB clean block, bulk initializing the unlog table once is definitely faster then setting the unlog table 1k times for each copied object.

Reply:
The thing is that we may not copy an object at all in sticky immix. The object is allocated with `log bit === 0`, but when we trace it, we need to set its log bit to 1. However, I can try this: for objects that are copied, bulk set their log bit, for objects that are not copied, set the log bit per object.